experimental/dds/tree2/src/codec/parameterized-dds.md

```diff
@@ -0,0 +1,252 @@
+		runtime: IFluidDatastoreRuntime,
+		id?: string,
+		params?: Partial<SharedSomethingOptions>,
+	): SharedSomething {
+		// Uh oh--params unused!
+		return runtime.createChannel(id, SharedSomethingFactory.Type) as SharedSomething;
+	}
+
+	public static getFactory(params?: Partial<SharedSomethingOptions>): SharedSomethingFactory {
+		return new SharedSomethingFactory(params);
+	}
+}
+```
+
+This immediately uncovers one issue: since the static `create` method isn't in charge of constructing the DDS's factory,
+there's no place to plumb through the parameters.
+Even worse, the container author likely already wrote some code like this:
+
+```typescript
+const myDataObject = new DataObjectFactory("foo", MyDataObject, [
+	SharedSomething.getFactory({ formatVersion: 5 }),
+]);
+```
+
+If they later write `SharedSomething.create(runtime, { formatVersion: 4 })`, the create flow outlined above will instead use the
+factory created at data object registry construction time.
+
+There is another, more subtle issue with this strategy: how should one implement a `SharedSomethingFactory` which takes in some parameters?
+
+```typescript
+class SharedSomethingFactory {
+	public constructor(private readonly options: SharedSomethingOptions) {}
+
+	public load(
+		runtime: IFluidDataStoreRuntime,
+		id: string,
+		services: IChannelServices,
+		channelAttributes: Readonly<IChannelAttributes>,
+	): Promise<IChannel> {
+		const sharedSomething = new SharedSomething(
+			runtime,
+			id,
+			this.options /* uh oh: channelAttributes and this.options might disagree on snapshot format! */,
+		);
+		await sharedSomething.load(services);
+		return sharedSomething;
+	}
+
+	public create(runtime: IFluidDataStoreRuntime, id: string): IChannel {
+		const sharedSomething = new SharedSomething(document, id, this.options);
+		sharedSomething.initializeLocal();
+		return sharedSomething;
+	}
+}
+```
+
+If the channel's attributes clearly reflect its parameters (e.g. the snapshot format being used),
+then the resulting factory sometimes produces DDSes which don't match the requested format!
+
+There are essentially two ways to resolve this issue:
+
+1. Make `SharedSomethingFactory` have a parameterless constructor, and pass parameters all the way through the `create` flow (i.e. through `createChannel`, factory lookup, and factory's `create` method).
+2. Keep parameters in `SharedSomethingFactory`'s constructor, but change their semantics such that they only necessarily apply to newly created documents.
+
+Regardless of which option we pick, enough information to reify the DDS's parameters needs to be persisted in the attributes blob.
+
+In order to support data objects which can create multiple types of the same shared object with different parameters, option 2 would need to be accompanied by changes to make DDS registration take in named entries in `@fluidframework/aqueduct`, rather than assume the "key" for a DDS registry entry is that DDS's factory's `type` field.
```

Comment:
Might be easier to reason about some of this if you peel the aqueduct layer off - I found it was nice to remove that abstraction layer for considering deeper changes to initialization flows in #15932

Reply:
interestingly, it's the aqueduct layer that introduces self-naming factories. Lower layers take an `ISharedObjectRegistry`.

Reply:
The ISharedObjectRegistry is kind of a lie in practice though - the names it holds are what we use for lookup, but our SharedObjects ignore those names completely when creating a new object.  I tested this in my draft PR -- using some other name for the SharedCell factory fails because we still create with type `"https://graph.microsoft.com/types/cell"` in the attributes anyway, so on next load we get a channelFactoryNotRegisteredForGivenType.

Reply:
yeah, that's a fair point. I'd put that under "the attributes blob needs enough info to reify the right named entry," but you're right some of the base APIs or implementations might need to change to make that happen. i'll look into it

Reply:
... I see that's the same point you're getting at in https://github.com/microsoft/FluidFramework/pull/15999#discussion_r1230253747. Makes sense.

Reply:
Yeah - summarizeChannel currently just consults the channel itself about [what it thinks its attributes are](https://github.com/microsoft/FluidFramework/blob/main/packages/runtime/datastore/src/channelContext.ts#L106).

But there's probably an opportunity for [getAttachSummary](https://github.com/microsoft/FluidFramework/blob/main/packages/runtime/datastore/src/localChannelContext.ts#L131) to instead pass in the type (as according to the LocalChannelContext who [knows the real registered name here](https://github.com/microsoft/FluidFramework/blob/main/packages/runtime/datastore/src/localChannelContext.ts#L298)).